source/common/network/connection_impl.cc

```diff
@@ -190,13 +191,31 @@ void ConnectionImpl::onRead(uint64_t read_buffer_size) {
     return;
   }
 
-  if (read_buffer_size == 0) {
+  if (read_buffer_size == 0 && !read_end_stream_) {
     return;
   }
 
+  if (read_end_stream_) {
+    if (read_end_stream_raised_) {
+      // No further data can be delivered after end_stream
+      ASSERT(read_buffer_size == 0);
```

Comment:
Out of curiosity when does this happen? Might be worth some more comments on why this occurs (I can't think of it off the top of my head).

Reply:
But why do we keep calling read()? Are there cases in which ready ready keeps getting set?

Reply:
Neither @alyssawilk nor myself could convince ourselves that it couldn't happen, and we wanted to make a guarantee to filters that they wouldn't see duplicates.

Reply:
OK fair enough. Do you mind expanding the comment a bit more to say that we don't think this should happen but this is an extra paranoid guard? (Or something to that effect).